### PR TITLE
feat: Allow metadata filtering for fetch records endpoints

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

The purpose of this PR is to add support for metadata filtering to API endpoints fetching records:
- `/api/v1/datasets/:dataset_id/records`
- `/api/v1/me/datasets/:dataset_id/records`

by using the same metadata query pattern defined for the search endpoint 

To support this, the `SearchEngine` have been modify to allow searches without a query (just for filtering)

Refs #3748 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] New feature (non-breaking change which adds functionality)
- [x] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

Some local tests have been done.

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/server/apis/v1/handlers/datasets.py [hunk 1]
```diff
@@ -332,6 +372,36 @@ async def create_dataset_metadata_property(
...
+
+    # TODO: We should split API v1 into different FastAPI apps so we can customize error management.
+    # After mapping ValueError to 422 errors for API v1 then we can remove this try except.
+    try:
+        metadata_property = await datasets.create_metadata_property(db, dataset, metadata_prop_create)
+        return metadata_property
+    except ValueError as err:
+        raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(err))
+
+
```
Comment (gabrielmbmb) [inline src/argilla/server/apis/v1/handlers/datasets.py]: this is duplicate

File: src/argilla/server/apis/v1/handlers/datasets.py [hunk 2]
```diff
@@ -474,6 +542,47 @@ async def search_dataset_records(
     )
 
 
+async def _build_metadata_filters(
+    db: "AsyncSession", dataset: Dataset, parsed_metadata: List[MetadataParsedQueryParam]
+):
```
Comment (gabrielmbmb) [inline src/argilla/server/apis/v1/handlers/datasets.py]:
```suggestion
) -> List["MetadataFilter"]:
```

Comment (gabrielmbmb) [reply src/argilla/server/apis/v1/handlers/datasets.py]: will need to import `MetadataFilter`

File: src/argilla/server/apis/v1/handlers/datasets.py [hunk 3]
```diff
@@ -150,15 +160,33 @@ async def list_current_user_dataset_records(
...
+            user_response_status_filter=response_status_filter,
+            offset=offset,
+            limit=limit,
+            sort_by=[SortBy(field="inserted_at")],
+        )
+
+        record_ids = [response.record_id for response in search_responses.items]
+        records = await datasets.get_records_by_ids(
+            db=db, dataset_id=dataset_id, record_ids=record_ids, include=include
+        )
```
Comment (gabrielmbmb) [inline src/argilla/server/apis/v1/handlers/datasets.py]: maybe we could move this to a function, so it could be reused in both endpoints

**Review by gabrielmbmb — APPROVED** [review]

File: src/argilla/server/apis/v1/handlers/datasets.py (hunk 1 above)
Comment (frascuchon) [reply src/argilla/server/apis/v1/handlers/datasets.py]: Done (again) 🙃

File: src/argilla/server/apis/v1/handlers/datasets.py (hunk 3 above)
Comment (frascuchon) [reply src/argilla/server/apis/v1/handlers/datasets.py]: Done

**Pull request merged**